casper/src/main/resources/RevVault.rho

```diff
@@ -16,266 +16,262 @@
  */
 new
   RevVault, rs(`rho:registry:insertSigned:secp256k1`), uriOut,
-  rl(`rho:registry:lookup`), RevAddress(`rho:rev:address`), MakeMintCh, AuthKeyCh, EitherCh,
+  rl(`rho:registry:lookup`), RevAddress(`rho:rev:address`),
+  MakeMintCh, AuthKeyCh, EitherCh, TreeHashMapCh,
   _makeVault,
   _newVault,
-  _getOrCreate,
+  _create,
   _revVault,
   _transferTemplate,
   _depositTemplate
 in {
   rl!(`rho:rchain:makeMint`, *MakeMintCh) |
   rl!(`rho:rchain:authKey`, *AuthKeyCh) |
   rl!(`rho:lang:either`, *EitherCh) |
-  for(@(_, MakeMint) <- MakeMintCh) {
-    for (@(_, AuthKey) <- AuthKeyCh) {
-      for (@(_, Either) <- EitherCh) {
-        new mintCh, vaultMapStore, initVault in {
-
-          @MakeMint!(*mintCh) |
-          for (mint <- mintCh) {
-
-            contract initVault(name, @address, @initialAmount) = {
-              new purseCh in {
-                mint!("makePurse", initialAmount, *purseCh) |
-                for (purse <- purseCh) {
-                  _newVault!(*name, address, *purse)
-                }
-              }
-            } |
-
-            for (@"init", ret <- RevVault) {
-              ret!(*vaultMapStore, *initVault)
-            } |
-
-            contract RevVault(@"deployerAuthKey", deployerId, ret) = {
-
-              new RevAddress(`rho:rev:address`), DeployerIdOps(`rho:rchain:deployerId:ops`), revAddrCh, deployerPubKeyBytesCh in {
-                DeployerIdOps!("pubKeyBytes", *deployerId, *deployerPubKeyBytesCh) |
-                for (@deployerPubKeyBytes <- deployerPubKeyBytesCh) {
-
-                  RevAddress!("fromPublicKey", deployerPubKeyBytes, *revAddrCh) |
-                  for (@deployerRevAddress <- revAddrCh) {
-
-                    @AuthKey!("make", (*_revVault, deployerRevAddress), *ret)
-                  }
-                }
-              }
-            } |
-
-            contract RevVault(@"unforgeableAuthKey", unf, ret) = {
-
-              new RevAddress(`rho:rev:address`), revAddrCh in {
-                RevAddress!("fromUnforgeable", *unf, *revAddrCh) |
-                for (@unfRevAddress <- revAddrCh) {
-
-                  @AuthKey!("make", (*_revVault, unfRevAddress), *ret)
-                }
+  rl!(`rho:lang:treeHashMap`, *TreeHashMapCh) |
+  for (@(_, MakeMint) <- MakeMintCh;
+       @(_, AuthKey)  <- AuthKeyCh;
+       @(_, Either)   <- EitherCh;
+       TreeHashMap    <- TreeHashMapCh) {
+    // Each invocation of the RevVault contract uses a different mint.
+    new mintCh, vaultMapStore, initVault in {
+      // Initializes the vaultMapStore.
+      TreeHashMap!("init", 3, *vaultMapStore) |
```

Comment:
Should probably use more than a depth-3 tree for the vault map, though I suppose it could be changed later with an RCIP if there's too much contention.

Reply:
What depth do you suggest using?

Reply:
I'd make it at least 4.

Reply:
Would there be a significant performance hit if it was made to be depth-4 and this depth wasn't necessary? For example, compared to depth-3?

Reply:
I was operating under the assumption (which is silly in hindsight) that the vaultMapStore would be updated whenever there was a payment. 3 is plenty; even 1 would suffice for foreseeable usage levels.

Reply:
Based on the above, it looks like we should certainly make this a tunable parameter, with a reasonable default value that would give decent performance.

Reply:
I PoS situation with pending rewards is even simpler than with REV vaults. @pyrocto should it be enough to set depth to 1 also?
https://github.com/rchain/rchain/blob/46b9acfdd3e2706a099a79ce30dd7149af3c0538/casper/src/main/resources/PoS.rhox#L106-L108

Reply:
Sure.